Mathlib/CategoryTheory/Sites/EffectiveEpimorphic.lean

```diff
@@ -0,0 +1,417 @@
+  { desc := fun S => Hf.desc (S.ι.app ⟨Over.mk f, ⟨𝟙 _, by simp⟩⟩) <| by
+      intro Z g₁ g₂ h
+      let Y' : D := ⟨Over.mk f, 𝟙 _, by simp⟩
+      let Z' : D := ⟨Over.mk (g₁ ≫ f), g₁, rfl⟩
+      let g₁' : Z' ⟶ Y' := Over.homMk g₁
+      let g₂' : Z' ⟶ Y' := Over.homMk g₂ (by simp [h])
+      change F.map g₁' ≫ _ = F.map g₂' ≫ _
+      simp only [S.w]
+    fac := by
+      rintro S ⟨T,g,hT⟩
+      dsimp
+      nth_rewrite 1 [← hT, Category.assoc, Hf.fac]
+      let y : D := ⟨Over.mk f, 𝟙 _, by simp⟩
+      let x : D := ⟨Over.mk T.hom, g, hT⟩
+      let g' : x ⟶ y := Over.homMk g
+      change F.map g' ≫ _ = _
+      rw [S.w]
+      rfl
+    uniq := by
+      intro S m hm
+      dsimp
+      generalize_proofs h1 h2
+      apply Hf.uniq _ h2
+      exact hm ⟨Over.mk f, 𝟙 _, by simp⟩ }
+
+/--
+Implementation: This is a construction which will be used in the proof that
+the sieve generated by a single arrow is effective epimorphic if and only if
+the arrow is an effective epi.
+-/
+noncomputable
+def effectiveEpiStructOfIsColimit {X Y : C} (f : Y ⟶ X)
+    (Hf : IsColimit (Sieve.generateSingleton f : Presieve X).cocone) :
+    EffectiveEpiStruct f :=
+  let aux {W : C} (e : Y ⟶ W)
+    (h : ∀ {Z : C} (g₁ g₂ : Z ⟶ Y), g₁ ≫ f = g₂ ≫ f → g₁ ≫ e = g₂ ≫ e) :
+    Cocone (Sieve.generateSingleton f).arrows.diagram :=
+    { pt := W
+      ι := {
+        app := fun ⟨T,hT⟩ => hT.choose ≫ e
+        naturality := by
+          rintro ⟨A,hA⟩ ⟨B,hB⟩ (q : A ⟶ B)
+          dsimp ; simp only [← Category.assoc, Category.comp_id]
+          apply h
+          rw [Category.assoc, hB.choose_spec, hA.choose_spec, Over.w] } }
+  { desc := fun {W} e h => Hf.desc (aux e h)
+    fac := by
+      intro W e h
+      dsimp
+      have := Hf.fac (aux e h) ⟨Over.mk f, 𝟙 _, by simp⟩
+      dsimp at this ; rw [this] ; clear this
+      nth_rewrite 2 [← Category.id_comp e]
+      apply h
+      generalize_proofs hh
+      rw [hh.choose_spec, Category.id_comp]
+    uniq := by
+      intro W e h m hm
+      dsimp
+      apply Hf.uniq (aux e h)
+      rintro ⟨A,g,hA⟩
+      dsimp
+      nth_rewrite 1 [← hA, Category.assoc, hm]
+      apply h
+      generalize_proofs hh
+      rwa [hh.choose_spec] }
+
+theorem Sieve.effectiveEpimorphic_singleton {X Y : C} (f : Y ⟶ X) :
+    (Presieve.singleton f).EffectiveEpimorphic ↔ (EffectiveEpi f) := by
+  constructor
+  · intro (h : Nonempty _)
+    rw [Sieve.generateSingleton_eq] at h
+    constructor
+    apply Nonempty.map (effectiveEpiStructOfIsColimit _) h
+  · rintro ⟨h⟩
+    show Nonempty _
+    rw [Sieve.generateSingleton_eq]
+    apply Nonempty.map (isColimitOfEffectiveEpiStruct _) h
+
+/--
+The sieve of morphisms which factor through a morphism in a given family.
+This is equal to `Sieve.generate (Presieve.ofArrows X π)`, but has
+more convenient definitional properties.
+-/
+def Sieve.generateFamily {B : C} {α : Type _} (X : α → C) (π : (a : α) → (X a ⟶ B)) :
+    Sieve B where
+  arrows Y f := ∃ (a : α) (g : Y ⟶ X a), g ≫ π a = f
```

Comment:
I guess that what is going on here is that you need to define a `Presieve`, so a `Set` (of arrows) and this is a map to `Prop`, that you are giving with this `∃` term? Perhaps this deserves a couple of lines of explanation?

Reply:
Sure, I can use set-builder notation to make this more readable.